cpp/src/arrow/array/array_dict_test.cc

```diff
@@ -1428,19 +1429,22 @@ TEST(TestDictionary, IndicesArray) {
   ASSERT_OK(arr->indices()->ValidateFull());
 }
 
+void CheckDictionaryCompact(const std::shared_ptr<Array>& input,
+                            const std::shared_ptr<Array>& expected) {
+  const auto& input_ref = checked_cast<const DictionaryArray&>(*input);
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Array> actual, input_ref.Compact());
```

Comment:
Since we are at it, can we also validate the output?
```suggestion
  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Array> actual, input_ref.Compact());
  ASSERT_OK(actual->ValidateFull());
```